storage_controller/src/service.rs

```diff
@@ -1949,17 +1949,12 @@ impl Service {
 
         let mut cleanup = Vec::new();
         {
-            let mut locked = self.inner.write().unwrap();
+            let locked = self.inner.read().unwrap();
 
-            for (tenant_shard_id, observed_loc) in configs.tenant_shards {
-                let Some(tenant_shard) = locked.tenants.get_mut(&tenant_shard_id) else {
+            for (tenant_shard_id, _observed_loc) in configs.tenant_shards {
+                if !locked.tenants.contains_key(&tenant_shard_id) {
                     cleanup.push(tenant_shard_id);
-                    continue;
-                };
-                tenant_shard
```

Comment:
By not modifying `observed` here, we're assuming that it's already either correct or None, right?  Is that a true assumption?

Reply:
Indeed. I think it should always be correct, but we can mark any mismatches as `None` to err on the side of caution.

Reply:
Yeah, that feels like a nice anti-fragile thing to do:
- if they match, great, no need to interfere
- if they don't, then we shouldn't trust the state we see here (as it might reflect a change that didn't generate a compute notification yet), so we should None it

Reply:
Indeed. Done in https://github.com/neondatabase/neon/pull/11155/commits/9fc5a75b2ed8204b10e8905d51f2f9889f36b44e.